### PR TITLE
lan-mouse: init at 0.6.0

### PR DESCRIPTION
**Pull request by PedroHLC**
## Description of changes

Adds https://github.com/feschber/lan-mouse

Another promising KVM that should work with Wayland…

**EDIT**: Tested and it works between sway<->sway

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Previous non-nixpkgs attempts (this PR was not based on them):
- https://github.com/feschber/lan-mouse/issues/80
- https://github.com/feschber/lan-mouse/pull/82

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  src = fetchFromGitHub {
+    owner = "feschber";
+    repo = "lan-mouse";
+    rev = "v${version}";
+    hash = "sha256-98n0Y9oL/ll90NKHJC/25wkav9K+eVqrO7PlrJMoGmY=";
+  };
+
+  nativeBuildInputs = [
+    glib # needed in both {b,nativeB}uildInptus
+    pkg-config
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:25]: I'd expect to see `wrapGAppsHook4` as well

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+, fetchFromGitHub
+, lib
+, pkg-config
+, glib
+, gtk4
+, libadwaita
+, xorg
+, libX11 ? xorg.libX11
+, libXtst ? xorg.libXtst
+}:
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:11]:
```suggestion
, libX11
, libXtst
}:

```

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    outputHashes = {
+      "reis-0.1.0" = "sha256-sRZqm6QdmgqfkTjEENV8erQd+0RL5z1+qjdmY18W3bA=";
+    };
+  };
+
+  meta = with lib; {
+    description = "Mouse & keyboard sharing via LAN";
+    homepage = "https://github.com/feschber/lan-mouse";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ pedrohlc ];
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:47]: please set `meta.mainProgram`

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "reis-0.1.0" = "sha256-sRZqm6QdmgqfkTjEENV8erQd+0RL5z1+qjdmY18W3bA=";
+    };
+  };
+
+  meta = with lib; {
+    description = "Mouse & keyboard sharing via LAN";
```
Comment (superherointj) [inline pkgs/by-name/la/lan-mouse/package.nix:44]:
```suggestion
    description = "A software KVM switch for sharing a mouse and keyboard with multiple hosts through the network";
```

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    homepage = "https://github.com/feschber/lan-mouse";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ pedrohlc ];
+    platforms = lib.platforms.unix;
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:48]: upstream claims windows support as well

**Review by eclairevoyant — DISMISSED** [review]
Some feedback below.

Also, the program fails to start on my end. `strace` indicates that it can't find the applicable socket:

```
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0) = 10
connect(10, {sa_family=AF_UNIX, sun_path="/run/user/1000/lan-mouse-socket.sock"}, 39) = -1 ENOENT (No such file or directory)
close(10)
```

File: pkgs/by-name/la/lan-mouse/package.nix (hunk 5 above)
Comment (PedroHLC) [reply pkgs/by-name/la/lan-mouse/package.nix:48]: My bad, I thought that Cygwin/Mingw would fall under Unix as well.

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    outputHashes = {
+      "reis-0.1.0" = "sha256-sRZqm6QdmgqfkTjEENV8erQd+0RL5z1+qjdmY18W3bA=";
+    };
+  };
+
+  meta = {
+    description = "A software KVM switch for sharing a mouse and keyboard with multiple hosts through the network";
+    homepage = "https://github.com/feschber/lan-mouse";
+    license = lib.licenses.gpl3Only;
+    mainProgram = pname;
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:48]:
```suggestion
    mainProgram = "lan-mouse";
```
https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+, fetchFromGitHub
+, lib
+, pkg-config
+, glib
+, gtk4
+, libadwaita
+, libX11
+, libXtst
+, wrapGAppsHook4
+}:
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:11]: nit:
```suggestion
}:

```

**Comment by PedroHLC** [comment]
> Also, the program fails to start on my end. `strace` indicates that it can't find the applicable socket:

I'm testing it here with a simple `nix run .#` and with `nix build .#` over two systems running latst `nixos-unstable`. Is that in a service?

**Comment by eclairevoyant** [comment]
> > Also, the program fails to start on my end. `strace` indicates that it can't find the applicable socket:
>
> Is that in a service?

No, I'm running directly from the terminal. No GUI opens, and the output is as follows:

```
$ result/bin/lan-mouse
[2024-02-29T14:00:36Z WARN  lan_mouse::config] /home/user/.config/lan-mouse/config.toml: No such file or directory (os error 2)
[2024-02-29T14:00:36Z WARN  lan_mouse::config] Continuing without config file ...
[2024-02-29T14:00:36Z WARN  lan_mouse::config] /home/user/.config/lan-mouse/config.toml: No such file or directory (os error 2)
[2024-02-29T14:00:36Z WARN  lan_mouse::config] Continuing without config file ...
[2024-02-29T14:00:36Z INFO  lan_mouse] Press Ctrl+Alt+Shift+Super to release the mouse
[2024-02-29T14:00:36Z INFO  lan_mouse::consumer] using wlroots event consumer
[2024-02-29T14:00:36Z INFO  lan_mouse::producer] libei event producer not available: not implemented
[2024-02-29T14:00:36Z INFO  lan_mouse::producer] using layer-shell event producer
[2024-02-29T14:00:36Z ERROR lan_mouse] io error: No such file or directory (os error 2)
```

**Comment by PedroHLC** [comment]
> > Is that in a service?
> 
> No, I'm running directly from the terminal. No GUI opens, and the output is as follows:
> 
> And I tested this with the gtk wrapper.

Interesting, you're using the same consumers as I, but it works here, from terminal or not.

**Review by itslychee — APPROVED** [review]

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    wrapGAppsHook4
+  ];
+
+  buildInputs = [
+    glib
+    gtk4
+    libadwaita
+    libX11
+    libXtst
+  ];
```
Comment (eclairevoyant) [inline pkgs/by-name/la/lan-mouse/package.nix:36]:
```suggestion
  ]
  ++ lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.CoreGraphics;
```

to fix darwin build

**Review by eclairevoyant — COMMENTED** [review]
Must be a config issue on my end, I can run this fine on my Arch system but not NixOS.

One more change to pass CI:

File: pkgs/by-name/la/lan-mouse/package.nix [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "reis-0.1.0" = "sha256-sRZqm6QdmgqfkTjEENV8erQd+0RL5z1+qjdmY18W3bA=";
+    };
+  };
+
+  meta = {
+    description = "A software KVM switch for sharing a mouse and keyboard with multiple hosts through the network";
+    homepage = "https://github.com/feschber/lan-mouse";
+    license = lib.licenses.gpl3Only;
```
Comment (fabaff) [inline pkgs/by-name/la/lan-mouse/package.nix:51]: Will allow tools (e.g., `nix-update`) to add a link to the commit message/PR description.

```suggestion
    changelog = "https://github.com/feschber/lan-mouse/releases/tag/v${version}";
    license = lib.licenses.gpl3Only;
```

**Review by superherointj — APPROVED** [review]

**Comment by feschber** [comment]
@PedroHLC  Hi, heres the developer of lan-mouse. I'm not too familiar with nix. Do I open a pr to get the version updated?

**Comment by superherointj** [comment]
> Do I open a pr to get the version updated?

Yes.

**Comment by feschber** [comment]
#298010

**Pull request merged**